infrastructure/index.ts

```diff
@@ -267,20 +261,65 @@ crawlDirectory(
     webContentsRootPath,
     (filePath: string) => {
         const relativeFilePath = filePath.replace(webContentsRootPath + "/", "");
+
+        // If the file is a .html file that contains a meta refresh redirect, configure the bucket object to be a
+        // webpage redirect to make S3 return a 301 response when the object is requested. This is done to allow
+        // #anchors in URLs to flow across to the destination page, which browsers will do with 301 redirects, but won't
+        // do with meta refresh redirects.
+        // https://docs.aws.amazon.com/AmazonS3/latest/dev/how-to-page-redirect.html
+        const redirect = getMetaRefreshRedirect(filePath);
+
         const contentFile = new aws.s3.BucketObject(
             relativeFilePath,
             {
                 acl: "public-read",
                 key: relativeFilePath,
                 bucket: contentBucket,
-                source: new pulumi.asset.FileAsset(filePath),
-                contentType: getMimeType(filePath) || undefined,
+
+                // Use "/dev/null" for an empty object if this is a redirect, otherwise the actual file path.
+                source: new pulumi.asset.FileAsset(redirect ? "/dev/null" : filePath),
+
+                // Add `contentType` if this isn't a redirect.
+                ...!redirect && { contentType: getMimeType(filePath) },
```

Comment:
Personally this `...!redirect && { obj-literal }` is really hard to understand.

Why don't we just create the `aws.s3.BucketObject` in two separate parts of an if-statement. e.g.

```
// Create a new S3 object for most files, but HTML files that contain a redirect
// just create a stubbed out S3 object with the right metadata so they return a
// 301 redirect (instead of serving the HTML with a meta-redirect). This ensures
// the right HTTP code response is returned for search engines, as well as
// enables better support for URL anchors.
if (!redirect) { 
   const contentFile = new aws.S3.BucketObject(
        ...);
} else {
    const s3Redirect = new aws.s3.BucketObject(
        ...);
}
```

Reply:
OK. Done.